relations/src/shielder/deposit_and_merge.rs

```diff
@@ -71,45 +79,32 @@ mod relation {
         //------------------------------
         // Check the old note arguments.
         //------------------------------
-        let token_id = FpVar::new_input(ns!(cs, "token id"), || self.token_id())?;
-        let old_token_amount =
-            FpVar::new_witness(ns!(cs, "old token amount"), || self.old_token_amount())?;
-        let old_trapdoor = FpVar::new_witness(ns!(cs, "old trapdoor"), || self.old_trapdoor())?;
-        let old_nullifier = FpVar::new_input(ns!(cs, "old nullifier"), || self.old_nullifier())?;
-        let old_note = FpVar::new_witness(ns!(cs, "old note"), || self.old_note())?;
-
-        check_note(
-            &token_id,
-            &old_token_amount,
-            &old_trapdoor,
-            &old_nullifier,
-            &old_note,
-        )?;
+        let old_note = NoteVarBuilder::new(cs.clone())
+            .with_token_id(self.token_id(), Input)?
+            .with_token_amount(self.old_token_amount(), Witness)?
+            .with_trapdoor(self.old_trapdoor(), Witness)?
+            .with_nullifier(self.old_nullifier(), Input)?
+            .with_note(self.old_note(), Witness)?
+            .build()?;
 
         //------------------------------
         // Check the new note arguments.
         //------------------------------
-        let new_token_amount =
-            FpVar::new_witness(ns!(cs, "new token amount"), || self.new_token_amount())?;
-        let new_trapdoor = FpVar::new_witness(ns!(cs, "new trapdoor"), || self.new_trapdoor())?;
-        let new_nullifier = FpVar::new_witness(ns!(cs, "new nullifier"), || self.new_nullifier())?;
-        let new_note = FpVar::new_input(ns!(cs, "new note"), || self.new_note())?;
-
-        check_note(
-            &token_id,
-            &new_token_amount,
-            &new_trapdoor,
-            &new_nullifier,
-            &new_note,
-        )?;
+        let new_note = NoteVarBuilder::new(cs.clone())
+            .with_token_id_var(old_note.token_id.clone())
```

Comment:
`self.token_id` for consistency with above? Its just my personal preference, so feel free to ignore and resolve this

Reply:
not exactly I can do this - using `with_token_id(self.token_id(), ...)` will register another public input - this will break current code, as normally we provide `token_id` only once